### PR TITLE
feat: Upload a Reduced Fares file into an S3 bucket

### PR DESCRIPTION
**Pull request by arkadyan**
Asana ticket: [API Endpoint in Document Viewer for uploading a file](https://app.asana.com/0/1200240595613188/1200802304217405)

Upload a Reduced Fares file into an S3 bucket

Require an authorization token for API requests.
For now we will only have a single client, so we can just configure a
single authorized token.

### CODE REVIEW

**Comment by arkadyan** [comment]
@krisrjohnson21 Tagging you thinking you would probably like to see this work as well. I know this Phoenix stuff is still relatively new to you, but I'm happy to go through these changes in person if that seems helpful.

File: config/config.exs [hunk 1]
```diff
@@ -21,7 +21,8 @@ config :document_viewer,
   ],
   live_catalog?: true,
   redirect_http?: false,
-  cognito_group: "document-viewer"
+  cognito_group: "document-viewer",
+  upload_bucket: "mbta-ctd-reduced-fares-uploads"
```
[the remark this answers is not in the document]
Comment (arkadyan) [reply config/config.exs:25]: That's a good question. The initial reason that I think of for why we might want it to be env-specific configuration would be if we want to use different buckets in different environments. But the way we have it set up there's a single bucket and we will partition data within that bucket using per-environment folders. Are there other considerations I'm not thinking about for why we might want this as configuration?

[the remark this answers is not in the document]
Comment (paulswartz) [reply config/config.exs:25]: We go back&forth about whether things like S3 bucket names should be in repositories. But without a consistent standard, if all versions of the application are using this bucket than this is fine.

File: lib/document_viewer_web/ensure_api_auth_token.ex [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  Verify that the API request includes an Authorization header with a Bearer token.
+  """
+
+  import Plug.Conn
+
+  def init(opts) do
+    Keyword.put_new(
+      opts,
+      :api_auth_token,
+      System.get_env("API_AUTH_TOKEN", Application.get_env(:document_viewer, :api_auth_token))
```
[the remark this answers is not in the document]
Comment (paulswartz) [reply lib/document_viewer_web/ensure_api_auth_token.ex:12]: `runtime.exs` I mean

File: config/runtime.exs [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import Config
+
+config :document_viewer,
+  api_auth_token:
+    System.get_env("API_AUTH_TOKEN", Application.get_env(:document_viewer, :api_auth_token))
```
Comment (paulswartz) [inline config/runtime.exs:5]: Might be clearer as:
```suggestion
if token = System.get_env("API_AUTH_TOKEN") do
  config :document_viewer,
    api_auth_token: token
end
```

File: lib/document_viewer_web/ensure_api_auth_token.ex [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  @spec get_token(Plug.Conn.t()) :: {:ok, String.t()} | :error
+  defp get_token(conn) do
+    case get_req_header(conn, "authorization") do
+      ["Bearer " <> token] -> {:ok, token}
+      _res -> :error
+    end
+  end
+
+  @spec verify_token(String.t(), keyword()) :: :ok | :error
+  defp verify_token(token, api_auth_token: api_auth_token) do
```
Comment (paulswartz) [inline lib/document_viewer_web/ensure_api_auth_token.ex:37]: FWIW, this won't work if there are any other keywords in the list.

File: config/runtime.exs (hunk 3 above)
Comment (arkadyan) [reply config/runtime.exs]: Yeah that's nice, thanks.

File: lib/document_viewer_web/ensure_api_auth_token.ex (hunk 4 above)
Comment (arkadyan) [reply lib/document_viewer_web/ensure_api_auth_token.ex]: Oh right, that's no good, thanks. Cleaned this up.

**Review by paulswartz — APPROVED** [review]
🍰

**Pull request merged**